### PR TITLE
feat: Flex - Removed required field false positive for stop_id + added Foreign key violation for location_groups_id

### LINKED ISSUE #1749: Flex: Add location group schema and update stop time schema
### Describe the problem

We want to be able to support Flex as a part of #1721, we need to add a schema for the new file [location_groups.txt](https://gtfs.org/schedule/reference/#location_groupstxt) and [update the stop times schema](https://gtfs.org/schedule/reference/#stop_timestxt) to include new fields and modify stop_id from required to conditionally required.

### Proposed solution

New schema. Test by checking if:

- foreign_key_violation is triggered when location_group_id is defined in stop_times.txt but not in  location_groups.txt. Test with [modified-flex-feed-4226672-missing.zip](https://github.com/MobilityData/gtfs-validator/files/15266056/modified-flex-feed-4226672-missing.zip). I removed location_group_id 4226672 from location_groups.txt in https://data.trilliumtransit.com/gtfs/islandtransit-wa-us/islandtransit-wa-us--flex-v2-TEST.zip to test this.

- Check that missing_required_field no longer triggers when there's no stop_id in stop_times.txt AND a present `id` in `locations.geojson` OR `location_group_id` in `location_groups.txt`. Check with https://data.trilliumtransit.com/gtfs/islandtransit-wa-us/islandtransit-wa-us--flex-v2-TEST.zip

- Run performance tests on the implications of adding/checking additional fields in stop_times.txt 

- Update notice in #1810 to reflect requirements here for [`duplicate_geography_id`](https://docs.google.com/document/d/1QABoIeOdR6lAB7A6_X6WIbZwC7EnAIZvm_hRNwNQeoI/edit#heading=h.z8mjdmhjvvrb)

### Alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by jcpitre**
Closes #1749 

**Summary:**

> foreign_key_violation is triggered when location_group_id is defined in stop_times.txt but not in location_groups.txt. 

- Added a foreign key annotation to location_group_id in stopTimeSchema.

> Check that missing_required_field no longer triggers when there's no stop_id in stop_times.txt AND a present id in locations.geojson OR location_group_id in location_groups.txt. 

- Made the stop_id conditionally required. 
- Added missing fields in stoptimeSchema. 
- Added a validator that checks that no more than one of stop_id, location_goup_id and or stop_id is present
- the presence in locations.geojson is out of scope for this PR.

Also verified that feature `Zone-based Demand Responsive Transit` is present in the list of features for a dataset with Flex. We could not test this in #1776 until the @required was removed from stopId.

**Expected behavior:** 

Most of missing required fields related to stopId should disappear from reports.
Some other errors or warning that were masked by the missing required field notice should appear. Some of these will have to be corrected other issues. e.g. #1755 

Please make sure these boxes are checked before submitting your pull request - thanks!

- [x] Run the unit tests with `gradle test` to make sure you didn't break anything
- [ ] Add or update any needed [documentation](https://github.com/MobilityData/gtfs-validator/tree/master/docs) to the repo 
- [x] Format the title like "feat: [new feature short description]". Title must follow the Conventional Commit Specification(https://www.conventionalcommits.org/en/v1.0.0/).
- [x] Linked all relevant issues
- [ ] Include screenshot(s) showing how this pull request works and fixes the issue(s)

### CODE REVIEW

File: core/src/main/java/org/mobilitydata/gtfsvalidator/notice/ForbiddenGeographyIdNotice.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright 2020 Google LLC
```
Comment (davidgamez) [inline core/src/main/java/org/mobilitydata/gtfsvalidator/notice/ForbiddenGeographyIdNotice.java:2]: [picky]: 
```suggestion
 * Copyright 2024 MobilityData
```

File: main/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsStopTimeSchema.java [hunk 2]
```diff
@@ -67,6 +73,12 @@ public interface GtfsStopTimeSchema extends GtfsEntity {
   @CachedField
   String stopHeadsign();
 
+  @ConditionallyRequired
+  GtfsTime startPickupDropOffWindow();
+
+  @ConditionallyRequired
+  GtfsTime endPickupDropOffWindow();
```
Comment (davidgamez) [inline main/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsStopTimeSchema.java:80]: Is there any validator implementing the conditional required logic for these fields? If not, please remove the annotation and bring it back when the logic is implemented.

**Comment by jcpitre** [comment]
About the faliing check

- New errors:
   -  missing_trip_edge: Related to #1755
- Dropped errors:
   - missing_required_field: corrected by this PR
   -  trip_distance_exceeds_shape_distance: Not sure why it dropped TBD
  
- New warnings:
   -    TBD. One problem with the stop_id required field false positive fixed is that stop_times.txt is not considered unparseable any more. This could make some of the warnings appear because stop_times.txt is now parseable.

**Review by davidgamez — APPROVED** [review]

**Comment by emmambd** [comment]
@jcpitre These changes look to be expected and positive! Learnings from this: 

- I've added #1755 back into the 6.0 release since this is a false positive we'll need to correct beforehand. Outside the scope of this PR. 

- In general, I'm wondering about `gb-unknown-uk-aggregate-feed-gtfs-2014` in the performance tests. I'm assuming its performance issues are unrelated to the coding changes, but taking 7 extra minutes to run the validation report seems surprising. 

     Comparing the two most recent validation reports on prod of the Mobility Database: https://mobilitydatabase.org/feeds/mdb-2014 it looks like there's 34,000 fewer trips in the last dataset...not sure if we're using prod data in the acceptance tests. 

- Before these acceptance tests were run, I noticed only 1 feed in our analytics dashboard that had the `missing_required_field` notice because of stop_id and was unrelated to Flex data: https://mobilitydatabase.org/feeds/mdb-1000. It looks like it's a partial GTFS file (doesn't include all required files). This is a feed I would've expected to be affected by the new notice since `stop_id` is no longer included in `missing_required_field`. What am I missing?

**Comment by emmambd** [comment]
@jcpitre I also just want to clarify the merging order of these PRs. 

It looks like this can be merged independently on #1805 and #1810? Should adding the checks in the new validator that evaluates the unique id for geography in`locations.geojson` be a new issue or included in #1810?

**Comment by jcpitre** [comment]
> @jcpitre I also just want to clarify the merging order of these PRs.
> 
> It looks like this can be merged independently on #1805 and #1810? Should adding the checks in the new validator that evaluates the unique id for geography in`locations.geojson` be a new issue or included in #1810?
Do you mean this: https://github.com/MobilityData/gtfs-validator/blob/6f82f0368c95322d89939a62b680d4d207b9552d/main/src/main/java/org/mobilitydata/gtfsvalidator/validator/GtfsGeojsonFeatureUniqueLocationIdValidator.java
I don't see why it should have its own issue. It's kind of done in #1810 as a demonstration. Also now that location_groups.txt has been added I can just modify the validator to take it into account.

**Comment by emmambd** [comment]
Cool! @jcpitre so then it could be added as part of this PR?

**Comment by jcpitre** [comment]
> * In general, I'm wondering about `gb-unknown-uk-aggregate-feed-gtfs-2014` in the performance tests. I'm assuming its performance issues are unrelated to the coding changes, but taking 7 extra minutes to run the validation report seems surprising.
The stop_times in this datataset has nearly 42 million records. I added 5 new fields in the schema for that file. So at least for memory it's quite a chunk more, and it takes more to load. The new validator has to go through these 42 million records. Does it amount to 7 minutes? It could I guess.

**Comment by jcpitre** [comment]
With the new 16G heap the acceptance tests seem to run properly.

**Review by emmambd — APPROVED** [review]

**Pull request merged**